File: 701.insert_into_a_binary_search_tree/insert.go
// 701. Insret into a binary search tree
// Topics: 'Binary Search Tree', 'Tree', 'Binary Tree'

// You are given the root node of a binary search tree (BST) and a value to insert into the tree.
// Return the root node of the BST after the insertion.
// It is guaranteed that the new value does not exist in the original BST.

// Notice that there may exist multiple valid ways for the insertion, as long as the tree remains a BST after insertion. You can return any of them.

// Example 1:

// Input: root = [4,2,7,1,3], val = 5
// Output: [4,2,7,1,3,5]
// Explanation: Another accepted tree is:

// Example 2:

// Input: root = [40,20,60,10,30,50,70], val = 25
// Output: [40,20,60,10,30,50,70,null,null,25]

// Example 3:

// Input: root = [4,2,7,1,3,null,null,null,null,null,null], val = 5
// Output: [4,2,7,1,3,5]

// Constraints:

//     The number of nodes in the tree will be in the range [0, 104].
//     -108 <= Node.val <= 108
//     All the values Node.val are unique.
//     -108 <= val <= 108
//     It's guaranteed that val does not exist in the original BST.

package insertintoabinarysearchtree

func insertIntoBST(root *TreeNode, val int) *TreeNode {
	if root == nil {
		root = &TreeNode{Val: val}
		return root
	}

	if root.Val > val {
		root.Left = insertIntoBST(root.Left, val)
	} else {
		root.Right = insertIntoBST(root.Right, val)
	}

	// cur := root

	// for {
	// 	if cur.Val > val {
	// 		if cur.Left == nil {
	// 			cur.Left = &TreeNode{Val: val}
	// 			break
	// 		} else {
	// 			cur = cur.Left
	// 		}
	// 	} else {
	// 		if cur.Right == nil {
	// 			cur.Right = &TreeNode{Val: val}
	// 			break
	// 		} else {
	// 			cur = cur.Right
	// 		}
	// 	}
	// }

	return root
}

type TreeNode struct {
	Val   int
	Right *TreeNode
	Left  *TreeNode
}

File: 708.insert_into_a_sorted_circular_linked_list/insert.py
# 708. Insert into a Sorted Circular Linked List
# Topics: 'Linked List', 'Two Pointers'
# Level: 'Medium'

# Given a Circular Linked List node, which is sorted in non-descending order,
# write a function to insert a value insertVal into the list such that it remains
# a sorted circular list. The given node can be a reference to any single node in
# the list and may not necessarily be the smallest value in the circular list.
#
# If there are multiple suitable places for insertion, you may choose any place to
# insert the new value. After the insertion, the circular list should remain sorted.
#
# If the list is empty (i.e., the given node is null), you should create a new single
# circular list and return the reference to that single node. Otherwise, you should
# return the originally given node.

# Example 1:
# Input: head = [3,4,1], insertVal = 2
# Output: [3,4,1,2]

# Example 2:
# Input: head = [], insertVal = 1
# Output: [1]

# Example 3:
# Input: head = [1], insertVal = 0
# Output: [1,0]

# Constraints:
#     The number of nodes in the list is in the range [0, 5 * 10⁴].
#     -10⁶ <= Node.val, insertVal <= 10⁶

from typing import Optional


class Node:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next


class Solution:
    def insert(self, head: 'Optional[Node]', insertVal: int) -> 'Node':
        if not head:
            nNode = Node(insertVal)
            nNode.next = nNode
            return nNode

        biggest = head
        real_head = head
        cur = head
        while True:
            if biggest.val < cur.val:
                biggest = cur
            if real_head.val >= cur.val:
                real_head = cur
            cur = cur.next
            if cur == head:
                break

        real_tail = None
        cur = head
        while True:
            if cur.next == real_head:
                real_tail = cur
                break
            cur = cur.next

        if biggest.val > real_tail.val:
            real_tail = biggest
            real_head = biggest.next

        nNode = Node(insertVal)
        if insertVal < real_head.val or insertVal > real_tail.val:
            real_tail.next, nNode.next = nNode, real_tail.next
        else:
            cur = head
            while True:
                if cur.val <= insertVal <= cur.next.val:
                    cur.next, nNode.next = nNode, cur.next
                    break
                cur = cur.next

        return head
